x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -216,187 +229,175 @@ public WildcardFieldType clone() {
-                            // `ake` to complete the logic.
-                            tokens.add(lastUnusedToken);
+                        pathFragment = new StringBuilder();
+                    } else {
+                        if (i == 0) {
+                            // Prefix with start of string
+                            pathFragment.append(TOKEN_START_OR_END_CHAR);
                         }
-                        tokenizer.end();
-                        tokenizer.close();
-                    } catch (IOException ioe) {
-                        throw new ElasticsearchParseException("Error parsing wildcard query pattern fragment [" + fragment + "]");
+                        pathFragment.append(Character.toChars(codePoint));
                     }
                 }
+                if (pathFragment.length() > 0) {
+                    // Have a concrete end-of-string
+                    pathFragment.append(TOKEN_START_OR_END_CHAR);
+                    pathFragment.append(TOKEN_START_OR_END_CHAR);
+                    getNgramTokens(thisPathTokens, pathFragment.toString());
+                }
+                // Add one of several possible paths
+                if (thisPathTokens.size() > 0) {
+                    BooleanQuery onePath = createApproximationQuery(thisPathTokens);
+                    result.add(onePath, Occur.SHOULD);
+                }
             }
-
-            if (patternStructure.isMatchAll()) {
-                return new MatchAllDocsQuery();
+            BooleanQuery q = result.build();
+            if (q.clauses().size() == 0) {
+                return null;
             }
-            BooleanQuery approximation = createApproximationQuery(tokens);
-            if (approximation.clauses().size() > 1 || patternStructure.needsVerification()) {
-                BooleanQuery.Builder verifyingBuilder = new BooleanQuery.Builder();
-                verifyingBuilder.add(new BooleanClause(approximation, Occur.MUST));
-                Automaton automaton = WildcardQuery.toAutomaton(new Term(name(), wildcardPattern));
-                verifyingBuilder.add(new BooleanClause(new AutomatonQueryOnBinaryDv(name(), wildcardPattern, automaton), Occur.MUST));
-                return verifyingBuilder.build();
+            return q;
+        }
+        
+        protected void getNgramTokens(ArrayList<String> tokens, String fragment) {
+            // Break fragment into multiple Ngrams
+            TokenStream tokenizer = WILDCARD_ANALYZER.tokenStream(name(), fragment);
+            CharTermAttribute termAtt = tokenizer.addAttribute(CharTermAttribute.class);
+            String lastUnusedToken = null;
+            try {
+                tokenizer.reset();
+                boolean takeThis = true;
+                // minimise number of terms searched - eg for "12345" and 3grams we only need terms
+                // `123` and `345` - no need to search for 234. We take every other ngram.
```

Comment:
Should we skip 2 positions instead of 1 when there are enough characters ? `123456` should produce `123` and `456` rather than `123`, `345`, `456` ?

Reply:
The original thinking was this might help keep them more position-based by having the middle term act as a bridge

Reply:
I see, thanks.